pkg/bitmap/bitmap.go

```diff
@@ -0,0 +1,232 @@
+// Copyright 2021 The gVisor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package bitmap provides the implementation of bitmap.
+package bitmap
+
+import (
+	"math"
+	"math/bits"
+)
+
+// +stateify savable
+type Bitmap struct {
+	numOnes   int
+	bitBlock []uint64
+}
+
+// NewBitmap create a new empty Bitmap.
+func NewBitmap(size int) *Bitmap {
+	b := new(Bitmap)
+	bSize := size / 64
+	if size % 64 != 0 {
+		bSize = bSize + 1
+	}
+	b.bitBlock = make([]uint64, bSize, bSize)
+	return b
+}
+
+// IsEmpty verifiys if the Bitmap is empty.
+func (b *Bitmap) IsEmpty() bool {
+	return b.numOnes == 0
+}
+
+// Minimum return the smallest value in the Bitmap.
+func (b *Bitmap) Minimum() int32 {
+	for i := 0; i < len(b.bitBlock); i++ {
+		if w := b.bitBlock[i]; w != 0 {
+			r := bits.TrailingZeros64(w)
+			return int32(r + i*64)
+		}
+	}
+	return math.MaxInt32
+}
+
+// Maximum return the largest value in the Bitmap.
+func (b *Bitmap) Maximum() int32 {
+	for i := len(b.bitBlock); i > 0; i-- {
+		if w := b.bitBlock[i-1]; w != 0 {
+			r := bits.LeadingZeros64(w)
+			return int32((i-1)*64 + 63 - r)
+		}
+	}
+	return int32(0)
+}
+
+// Add add i to the Bitmap.
+func (b *Bitmap) Add(i int32) {
```

Comment:
All input parameters should be of type `uint32` instead of `int32`. It automatically highlights the precondition to the callers that they can not Add negative numbers. This change should be made in all functions. (Maximum(), Minimum(), ...)

Because a BitMap can only hold unsigned integers.

Reply:
> Because a BitMap can only hold unsigned integers.

Yes, I thought about this, but in order to sync the type of FD in `pkg/sentry/kernel/fd_table.go`, I use int32 here.
Otherwise, there would be lots of types conversion.

Reply:
I understand your point, but `bitmap` is its own package. It seems a little off to me that we bend it to fit one of its use cases which is FD tables. The implementation will quickly fall apart if negative integers are passed.

Maybe @ianlewis can also weigh in on this.